### PR TITLE
[SDK] Support streaming in eager mode

### PR DESCRIPTION
**Pull request by D-W-**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

This pull request involves changes across multiple files in the `src/promptflow` directory, primarily focusing on improving generator output handling and testing. The main changes include the introduction of generator output handling in `test_submitter.py` and `_script_executor.py`, addition of new test cases for generator outputs, and modification of existing methods to support generator outputs.

Generator output handling:

* [`src/promptflow/promptflow/_sdk/_submitter/test_submitter.py`](diffhunk://#diff-d9dd2c09a149b11eb54626edf065287eeb7b1a28e39719b8dd95377770f64138L474-R474): The `flow_test` method has been simplified to always call `_get_generator_outputs` on `line_result.output` [[1]](diffhunk://#diff-d9dd2c09a149b11eb54626edf065287eeb7b1a28e39719b8dd95377770f64138L474-R474). The `_get_generator_outputs` method has been expanded to handle both dictionary and generator type outputs [[2]](diffhunk://#diff-d9dd2c09a149b11eb54626edf065287eeb7b1a28e39719b8dd95377770f64138L588-R591).
* [`src/promptflow/promptflow/executor/_script_executor.py`](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5R6): Imported `GeneratorType` from the `types` module [[1]](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5R6). Updated the `exec_line` and `_exec_line` methods to accept a new `allow_generator_output` parameter [[2]](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5R50-R66). Depending on the value of `allow_generator_output`, the output is either stringified or left as is [[3]](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5R86). Added a new `_stringify_generator_output` method to convert generator output to a string [[4]](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5R104-R110).

Test case additions and modifications:

* [`src/promptflow/tests/sdk_cli_test/conftest.py`](diffhunk://#diff-9dced3762e0027d9b6b131576702c5290a82d72c20b2a828c67aa682d31d8e40R234-R238): Added a new fixture `stream_output` for testing generator outputs.
* [`src/promptflow/tests/sdk_cli_test/e2etests/test_flow_serve.py`](diffhunk://#diff-8f175978a83627600049f94831e06a3c9e3307cbde190fe5fbb187571ac514baR502-R547): Added a new test case `test_eager_flow_stream_output` to test different response types for generator outputs.
* [`src/promptflow/tests/sdk_cli_test/e2etests/test_flow_test.py`](diffhunk://#diff-acf3c2285ac2a7051d59e3a69abe02224e2d6d5d69e3a0aabd2afd9311f87657R371-R391): Added two new test cases `test_eager_flow_stream_output` and `test_stream_output_with_builtin_llm` to test generator outputs in eager flows and with built-in language models respectively.

Additional changes:

* [`src/promptflow/promptflow/_sdk/operations/_flow_operations.py`](diffhunk://#diff-afdd40a5d0519512dcf9be48bd46c4caaa2291b808687de77896989af63f47e4R183): A comment has been added in the `_test` method to indicate that `is_chat_flow` is set to True to allow generator output.
* [`src/promptflow/tests/test_configs/eager_flows/builtin_llm/builtin_call.py`](diffhunk://#diff-eef3589cf12002451966c15d6d2073d10505d4f4e1baa52076b3d114949f0355R1-R54): The `flow_entry` function has been updated to pass a `stream` parameter to the `chat` function.
* [`src/promptflow/tests/test_configs/eager_flows/stream_output/stream_output.py`](diffhunk://#diff-e5c1c37d4206be7c0ccc24a92f53b3bc605dad4660f46ba10a301fcd8579c0f3R1-R8): A new Python file has been added that defines a generator function `my_flow`.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_submitter/test_submitter.py [hunk 1]
```diff
@@ -585,4 +582,10 @@ def _raise_error_when_test_failed(test_result, show_trace=False):
     @staticmethod
     def _get_generator_outputs(outputs):
         outputs = outputs or {}
-        return {key: outputs for key, output in outputs.items() if isinstance(output, GeneratorType)}
+        if isinstance(outputs, dict):
+            generator_outputs = {key: outputs for key, output in outputs.items() if isinstance(output, GeneratorType)}
+            if generator_outputs:
+                logger.info(f"Some streaming outputs in the result, {generator_outputs.keys()}")
+        elif isinstance(outputs, GeneratorType):
+            logger.info("Streaming output in the result.")
```
Comment (thy09) [inline src/promptflow/promptflow/_sdk/_submitter/test_submitter.py:590]: We may refine the message to be more clear?

File: src/promptflow/promptflow/_sdk/_submitter/test_submitter.py [hunk 2]
```diff
@@ -584,5 +582,9 @@ def _raise_error_when_test_failed(test_result, show_trace=False):
 
     @staticmethod
     def _get_generator_outputs(outputs):
-        outputs = outputs or {}
-        return {key: outputs for key, output in outputs.items() if isinstance(output, GeneratorType)}
+        # covert output to dict to unify the log
+        outputs = convert_eager_flow_output_to_dict(outputs)
+        if isinstance(outputs, dict):
+            generator_outputs = {key: outputs for key, output in outputs.items() if isinstance(output, GeneratorType)}
```
Comment (lumoslnt) [inline src/promptflow/promptflow/_sdk/_submitter/test_submitter.py:588]: A typo here? `key: outputs` -> `key: output`

Comment (D-W-) [reply src/promptflow/promptflow/_sdk/_submitter/test_submitter.py:588]: Good catch!

**Review by thy09 — APPROVED** [review]

**Review by lumoslnt — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

File: src/promptflow/tests/test_configs/eager_flows/multiple_stream_outputs_dataclass/multiple_stream_outputs_dataclass.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+@dataclass
+class MyOutput:
+    output1: str
+    output2: str
+
+
+def my_flow(input_val: str = "gpt") -> MyOutput:
+    generator1 = (i for i in range(10))
+    generator2 = (i for i in range(10))
+    return MyOutput(
```
Comment (crazygao) [inline src/promptflow/tests/test_configs/eager_flows/multiple_stream_outputs_dataclass/multiple_stream_outputs_dataclass.py:15]: yield

**Review by crazygao — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**